test/e2e/volumes.go

```diff
@@ -0,0 +1,360 @@
+	expectNoError(err, "Cannot locate the server pod %v: %v", serverPod.Name, err)
+
+	By("sleeping a bit to give the server time to start")
+	time.Sleep(20 * time.Second)
+	return pod
+}
+
+// Clean both server and client pods.
+func volumeTestCleanup(client *client.Client, config VolumeTestConfig) {
+	By(fmt.Sprint("cleaning the environment after ", config.prefix))
+
+	defer GinkgoRecover()
+
+	podClient := client.Pods(config.namespace)
+
+	// ignore all errors, the pods may not be even created
+	podClient.Delete(config.prefix + "-client")
+	podClient.Delete(config.prefix + "-server")
+
+	if config.namespace != "default" {
+		client.Namespaces().Delete(config.namespace)
+	}
+}
+
+// Start a client pod using given VolumeSource (exported by startVolumeServer())
+// and check that the pod sees the data from the server pod.
+func testVolumeClient(client *client.Client, config VolumeTestConfig, volume api.VolumeSource, expectedContent string) {
+	By(fmt.Sprint("starting ", config.prefix, " client"))
+	podClient := client.Pods(config.namespace)
+
+	clientPod := &api.Pod{
+		TypeMeta: api.TypeMeta{
+			Kind:       "Pod",
+			APIVersion: "v1beta3",
+		},
+		ObjectMeta: api.ObjectMeta{
+			Name: config.prefix + "-client",
+			Labels: map[string]string{
+				"role": config.prefix + "-client",
+			},
+		},
+		Spec: api.PodSpec{
+			Containers: []api.Container{
+				{
+					Name:       config.prefix + "-client",
+					Image:      "gcr.io/google_containers/nginx:1.7.9",
+					Privileged: false,
+					Ports: []api.ContainerPort{
+						{
+							Name:          "web",
+							ContainerPort: 80,
+							Protocol:      api.ProtocolTCP,
+						},
+					},
+					VolumeMounts: []api.VolumeMount{
+						{
+							Name:      config.prefix + "-volume",
+							MountPath: "/usr/share/nginx/html",
+						},
+					},
+				},
+			},
+			Volumes: []api.Volume{
+				{
+					Name:         config.prefix + "-volume",
+					VolumeSource: volume,
+				},
+			},
+		},
+	}
+	if _, err := podClient.Create(clientPod); err != nil {
+		Failf("Failed to create %s pod: %v", clientPod.Name, err)
+	}
+	expectNoError(waitForPodRunningInNamespace(client, clientPod.Name, config.namespace))
+
+	By("reading a web page from the client")
+	body, err := client.Get().
+		Namespace(config.namespace).
+		Prefix("proxy").
+		Resource("pods").
+		Name(clientPod.Name).
+		DoRaw()
+	expectNoError(err, "Cannot read web page: %v", err)
+	Logf("body: %v", string(body))
+
+	By("checking the page content")
+	Expect(body).To(ContainSubstring(expectedContent))
+}
+
+////////////////////////////////////////////////////////////////////////////////
+// NFS
+////////////////////////////////////////////////////////////////////////////////
+var _ = Describe("volumes-NFS", func() {
+	clean := true // If 'false', the test won't clear its namespace (and pods) upon completion. Useful for debugging.
+
+	// filled in BeforeEach
+	var c *client.Client
+	var namespace *api.Namespace
+
+	BeforeEach(func() {
```

Comment:
It would be good to factor the body of this into a separate function in this file, as it's repeated verbatim in the other test.
Of course it's also repeated almost verbatim in may other test files, but we can fix that in a separate PR.